sstables/sstables.cc

```diff
@@ -106,12 +106,19 @@ read_monitor_generator& default_read_monitor_generator() {
     return noop_read_monitor_generator;
 }
 
-static future<file> open_sstable_component_file_non_checked(std::string_view name, open_flags flags, file_open_options options,
+future<file> sstable::open_sstable_component_file_non_checked(std::string_view name, open_flags flags, file_open_options options,
         bool check_integrity) noexcept {
-    if (flags != open_flags::ro && check_integrity) {
-        return open_integrity_checked_file_dma(name, flags, options);
-    }
-    return open_file_dma(name, flags, options);
+    return std::visit(overloaded_functor {
+        [&] (const data_dictionary::storage_options::local&) {
+            if (flags != open_flags::ro && check_integrity) {
+                return open_integrity_checked_file_dma(name, flags, options);
+            }
+            return open_file_dma(name, flags, options);
+        },
+        [&] (const data_dictionary::storage_options::s3&) {
+            throw std::runtime_error("Support for S3 is not implemented yet");
+        }
```

Comment:
I'd suggest abstracting underlying storage, as sstable layer doesn't need to know about each storage option, nor validate it. Validation should be pushed to a layer above.

e.g.:
```
struct underlying_storage  {
    virtual future<file> open_file_dma(...);
    ...
};
```

Reply:
What do you mean by validating? In this change, I simply moved the previous code to the "::local" branch. The integrity check boolean is also part of the API that the sstable layer needs to pass, so I think it's this function that needs to be abstracted out. Right now it's done with variants, and I'm not sure what would be the benefit from using a virtual interface, since we statically know all the options. In which layer would you rather see this code? Since this std::visit needs to be ultimately called somewhere

Reply:
I'd like to avoid sstable layer being aware of data_dictionary::storage_options which is a concept that belongs to a higher layer. That's why I suggested the abstraction. The underlying storage itself is a layer underneath the sstable layer, but data_dictionary::storage_options is in a layer above.

To move this forward, I think we can move the visitor into a free function that sits in sstables.cc, then we can revisit it later.

By validation, I was referring to "Support for S3 is not implemented yet", this error should probably be triggered in a layer above, when user is trying to create a keyspace with S3 as underlying storage. But I understand this is just a stub. So I am fine.